### PR TITLE
[ULTRA WIP] Modular/template engines

### PR DESCRIPTION
**Pull request by Shadowlight213**
Basically, the Idea is to have the engine setup for engineering be a template which can be loaded in either randomly at roundstart, or possibly in the future the CE could choose somehow.
Anyways, here's my horribly mapped SM engine setup, squeezed into the space the Singularity was occupying.
sm

I haven't moved the SM setup to its own template file yet, because I'm still tweaking it. When I do, I'll either have the map default to sing setup, or have the area be empty.
TODO:
- [X] Make a supermatter setup that functions in the area occupied by the Singularity setup
- [ ] Get testmerge feedback on the setup, and make tweaks.
- [ ] Move everything to templates
- [ ] Get code to make it randomly select and load a template at roundstart.

### CODE REVIEW

**Comment by TechnoAlchemisto** [comment]
Doesn't random generation kind of force the players into doing a certain engine? 

Limiting round flexibility is probably a bad idea, even if the CE gets to choose.

**Comment by Iamgoofball** [comment]
:+1:

**Comment by Jalleo** [comment]
@TechnoAlchemisto hopefully this will encourage people to learn how each type of engine works.

**Comment by pudl** [comment]
I would rather engineering be remapped to accommodate for every engine type instead of being randomly generated. This will probably make engineering a hell of a lot harder to modify and remap at later points.

**Comment by PKPenguin321** [comment]
yeah, this is a neat idea but i'm not so sure about the implementation either.

**Comment by MMMiracles** [comment]
remapping engineering to fit every engine will just lead to people setting up the engine that easiest to do and maintain (tesla), like they already do.

randomly choosing an engine for them to work on actually gets people to work with other engines and if they genuinely don't feel competent enough to set up another engine, they can just do solars.

**Comment by ChemicalRascal** [comment]
I feel like doing something along the lines of having the shard randomly become the engine for a round or whatever needs to provide engineering with certain concessions -- namely, pipes. Otherwise, especially on lowpop, you'll have engineers who are really behind the eight-ball in setting it up.

(Which is to say they'll set up solars instead if they're above a certain threshold of competency and those that aren't will probably try to break into atmospherics, fail, and the station will run out of power.)

(Also maybe someone should write a super matter guide on the wiki before this gets mainlined but IDK)

**Comment by nullbear** [comment]
@TechnoAlchemisto 

> Doesn't random generation kind of force the players into doing a certain engine?

Doesn't having the singulo/tesla as the only roundstart engine kind of force players into doing a certain engine too tho?

The antimatter engine was also just recently added, but is otherwise impossible to use as a primary engine without admin intervention or a custom map, and as such, nobody will ever really get to use it.

As more different and interesting engines are added, as they likely will be (and have had plans to be added for ages) this is an extremely supportive method to accommodate them and introduce them to the players.

As for it having poor maintainability / adding difficulty to mappers. This should only affect maps that have templates built to support it. It would have the 'default' engine (depending on the map) and if alternative templates existed, it could optionally choose from one of them.

**Comment by nullbear** [comment]
That setup is hideous tho tbh and it still includes the PA, which is useless for the SM.

Incorporate the PA chamber into the template shape/size. Keep in mind that you CAN use space tiles as part of the template too?? it doesnt need to be a huge empty room.

**Comment by Anonmare** [comment]
Wiki needs a supermatter guide to be honest, one that's relevant to our codebase. Pinching bits off of Bay is degrading to say the least.

**Comment by TechnoAlchemisto** [comment]
Forcing players to pick one is much worse than forcing them to pick two. If the people who incorporated the supermatter and antimatter engines actually took effort to make them feasible to set up roundstart you would see them being used rather than putting them behind cargo to rot.

**Comment by Iamgoofball** [comment]
@TechnoAlchemisto and where are we going to pull this magic space and design to put all the engines on ALL SEVEN MAPS AT ONCE out of our ass from?

**Comment by TechnoAlchemisto** [comment]
all of the engines don't need to be on all the maps, but giving people the capability to only do one engine per round is bad, and really railroads engineering. 

just having a crate with rad collectors and a super matter shard in engineering would let people set up a SM engine without going through cargo.

**Comment by Iamgoofball** [comment]
@TechnoAlchemisto but they aren't

people don't do stuff that requires assembly unless they're really dedicated

you should know this by now, players are lazy

**Comment by nullbear** [comment]
Setting up the PA engines are just as time consuming as setting up the SM or AM. Except you start with the PA engines ALREADY set up. 

Under your logic, you would either have SM, AM, AND PA engine's 90% set up at round start. (Allowing for any or ALL engines to be active at round-start, at the cost of taking as much space as the PA engine does on itself.) 

Or you would have all engines 0% set up at round start for 'perfectly balanced choice' Which makes NO engines feasible for roundstart power. (where most people would pick PA anyways because they are more likely to know how to set it up from scratch, because it's what they're more used to.)

This is a perfect compromise, because it allows for the engines to be learned equally. 

Your typical player can't set up any engine from scratch because time constraints + inexperience + never having seen it to begin with. Hell even if the player does know how + there is enough time to do it, there's no reason to because the PA is already set up by that point in time.

The tesla + singulo count as one engine: **The PA**. The only difference in their setup is coils vs. collectors and whether or not you leave the PA on max.

We have four realistically reliable sources of energy right now: PA, SM, AM, and Solars.

**Comment by nullbear** [comment]
> all of the engines don't need to be on all the maps, but giving people the capability to only do one engine per round is bad, and really railroads engineering.

As opposed to what they have now? 

The only alternative to this would be to give engineering a huge empty space, and access to the materials they need to pick any engine they like. But the issue is that they have to build an engine before they run out of power, and even with the PA, it's unlikely that most engineers will be capable of doing that.

**Comment by ChemicalRascal** [comment]
Not to mention how terrible that would make lowpop --- You'd simply never see engines assembled. Ever. A lone engineer isn't going to be able to set it up in a reasonable length of time.

**Comment by nullbear** [comment]
Having ONE assembled engine + a disassembled engine is the same as having only one engine, because it's unrealistic to assemble an engine from scratch at round-start. And even MORE unrealistic to disassemble an engine BEFORE that.

Having a different possible (but still only one) assembled engine at round-start + a disassembled engine is the best compromise because you get to experience the different engines in a feasible manner, while still allowing engineers to partake in the bonus-project of building a second engine.

**Comment by Iamgoofball** [comment]
it's why I told the people who said "lol just put the tesla coils in storage people'll use it :^)" to fuck off

no one's going to use it you're just trying to bullshit your way into getting it off the map

**Comment by Bawhoppen** [comment]
:+1:

**Comment by Shadowlight213** [comment]
I've sorta lost interest in this for the moment. I might finish it later, but IDK.

**Comment by Iamgoofball** [comment]
Fuck

**Comment by ChemicalRascal** [comment]
Yeah, 's kind of a shame, this is an interesting idea IMO.

**Comment by Shadowlight213** [comment]
It's mainly mapping wise. I don't really feel much interest in continuing to work on mapping it and making it better. I'm fairly certain the current design wouldn't be accepted.
It's even not so much mapping itself, as figuring out how I need to design it.
I want power to come from the TEG, since just having collectors make the power doesn't leave much room for engineers to experiment with various mixes for better power generation.
However, I think that maybe a second engine is needed, which would require a redesign of the pipe layout, or an overhaul of the existing teg engine code.
Lighting is a pain, since It's a large open area.
Plus I really would like to have the PA room become a monitoring room with the sm module loaded.

Tldr: I can either just leave the map design as it is, and maybe try to look into how stuff needs to he loaded, or I'd need to redesign box engineering entirely to make it more friendly to having alternative engines loaded, mainly so I have more space to work with.
